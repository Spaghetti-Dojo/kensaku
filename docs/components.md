---
title: Components
layout: post
nav_order: 2
---

# Components

The Wp Entities Search provides a set of components you can use to build your search UI.

We have to distinguish between two types of components:

- **Base Components**: These are the basic building blocks of the search UI. They are the smallest components and are
  used to build more complex components.
- **Composite Components**: These are more complex components that are built using the base ones.

The _Composite Component_ does not have a direct dependency on the _Base Components_, indeed a composite component
encapsulate the logic
that deal with the state management of the data and pass it down to the base components in form of properties via
the `children` function.

In a nutshell, a _Composite Component_ is configured from the external via props, dialog with the Rest API and pass the
data down to the _Base Components_.

## Base Components

- `PluralSelectControl` - A select control that allows multiple selections.
- `SingleSelectControl` - A select control that allows a single selection.
- `SearchControl` - An input control that allows the user to enter a search query.
- `ToggleControl` - A multi selection control that allows the user to toggle between multiple options.
- `RadioControl` - A single selection control that allows the user to select a single option.

## Composite Components

- `CompositeEntitiesByKind` - A composite component that displays a list of entities by kind. In this context _kind_
  is `post` or `term` and _entity_ is the entity type of the content e.g. a `page` or a `category` term.

## Composite Entities by Kind

The `CompositeEntitiesByKind` is a _generic_ component that can be used to display a list of entities by kind. It acts
as a controller that fetches the data from the Rest API using a given function and pass it down to one or more _Base
Components_.
It is up to you how to consume the data given to the children. Note, even if you consume just one part of the given
data,
the component will fetch all the data necessary to render the UI.

This component is intended to fulfill a scenario where the user is looking for some entities belonging to a certain
_kind_,
for instance, selecting the `page` _kind_ will return a list of the available pages and to filter the result further it
is
possible to use a search function in conjunction with a `search` field.

An example of its usage is:

```jsx
import { CompositeEntitiesByKind } from 'wp-entities-search';

export function MyComponent(props) {
    const entities = {
        value: new Set([13, 24, 55]),
        onChange: (value) => {
            // Do something with the selected entities
            // Notice, here you'll receive the value as a string Set.
        }
    };

    const kind = {
        value: new Set(['page']),
        options: [
            { label: 'Pages', value: 'page' },
            { label: 'Posts', value: 'post' },
        ],
        onChange={(value) => {
            // Do something with the selected kind
        }}
    };

    const searchEntities = async (phrase, kind, queryArguments) => convertEntitiesToControlOptions(
        await searchEntities('post', kind, phrase, queryArguments),
        'title',
        'id'
    );

    return <CompositeEntitiesByKind
        searchEntities={searchEntities}
        entities={entities}
        kind={kind}
    >
        {(entities, kind, search) => (
            <>
                <ToggleControl
                    value={kind.value}
                    options={kind.options}
                    onChange={kind.onChange}
                />
                <SearchControl
                    onChange={search.onChange}
                />
                <ToggleControl
                    value={entities.value}
                    onChange={entities.onChange}
                />
            </>
        )}
    </CompositeEntitiesByKind>;
}
```

In the example above we are passing a `searchEntities` function that will be used to fetch the entities from the Rest
API. This function is documented in the [api](./api.md) section.

What's important to know is that the `queryArguments` do not include the `_fields` property because it's not a concern
of the component decide which fields to retrieve, it's up to the consumer to decide which fields to retrieve. Anyhow,
you can still override the given configuration, but you have to be aware this might produce an incorrect result if not
done correctly. The component might pass arguments necessary to guarantee the consistency of the Set of the entities and
the kind.

In the following example we require `title` and `slug` fields to be retrieved from the Rest API.

```typescript
 async (phrase, kind, queryArguments) => {
    const fields = ['title', 'slug'];
    return convertEntitiesToControlOptions(
        await searchEntities('post', kind, phrase, {...queryArguments, fields}),
        ...fields
    )
 }
```

The `entities` and `kind` properties are the initial configuration, they'll change depending on what happen within the
respective `onChange` functions.

The `children` function will receive the `entities`, `kind` and `search` properties; notice the `search` is a function,
and it's up to the `SearchControl` to maintain the status of the value, the composite does hold the search value
internally, but it does not share it outside.

You're not forced to consume all the properties given to the `children`, you can remove the `SearchControl`
and let the user select only the retrieved entities. Moreover, you can also not have the Kind control at all and just
allow the search for one single kind.

In the example below we only allow to select the entities belonging to the `page` or `landing-page` kind not permitting
the user to switch between them.

```jsx
import { CompositeEntitiesByKind } from 'wp-entities-search';

export function MyComponent(props) {
    const entities = {
        value: new Set([13, 24, 55]),
        onChange: (value) => {
            // Do something with the selected entities
        }
    };

    const kind = {
        value: new Set(['page', 'landing-page']),
        options: Set([]),
        onChange={() => {}}
    };

    const searchEntities = async (phrase, kind, queryArguments) => convertEntitiesToControlOptions(
        await searchEntities('post', kind, phrase, queryArguments),
        'title',
        'id'
    );

    return <CompositeEntitiesByKind
        searchEntities={searchEntities}
        entities={entities}
        kind={kind}
    >
        {(entities, _, search) => (
            <>
                <SearchControl
                    onChange={search.onChange}
                />
                <ToggleControl
                    value={entities.value}
                    options={entities.options}
                    onChange={entities.onChange}
                />
            </>
        )}
    </CompositeEntitiesByKind>;
}
```

Obviously depending on what you want to achieve you can use different _Base Components_ or create new ones, as mentioned
above the package comes with a set of _Base Components_ that can be used out of the box.

## About Singular Base Components

The _Composite Component_ always give a collection of Entities and Kind even though you are consuming a Single* _Base
Component_.

The Singular Components always get a single value, therefore you have to consider to extract the first element out of
the `Set`.

```jsx
import { CompositeEntitiesByKind } from 'wp-entities-search';

export function MyComponent(props) {
    return <CompositeEntitiesByKind
        /* ... */
    >
        {(entities, kind, search) => (
            <>
                <RadioControl
                    value={kind.value.first()}
                    options={kind.options}
                    onChange={(value) => kind.onChange(new Set([value]))}
                />
                /* ... */
            </>
        )}
    </CompositeEntitiesByKind>;
}
```
