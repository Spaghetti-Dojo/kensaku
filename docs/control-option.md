---
title: Control Option
layout: post
nav_order: 5
---

# Control Option

The `ControlOption` is the core Value Object of the project. It holds the information about the entities, and it
is carried around the business logic and the presentational layer.

The Control Option mainly contain the following properties:

- `label` (String): The label of the control option.
- `value` (mixed): The value of the control option.

An additional data structure is part of the `EnrichedControlOption` interface implemented by `ControlOption`. This DS is
not consumed within any of the _Base_ or _Composite_ components of the project, it is there to permit the third parties
to customize the UI.

The `extra` property is of type `Kensaku.Record`; it can hold a heterogeneous set of properties, we do not check
against values or types, this is up to the consumer.

## About the Value

The `value` is a generic type to give the possibility to customize the shape of the value since we allow the consumer to
implement they own components, and therefore we do not want to pose any restriction.

Nonetheless, the possible types the `value` can assume within the project Components is `string | number`;
see `Kensaku.Value` type for more info.

## How to use the Extra Property

Let's say you're using the `CompositeEntitiesByKind` component, and you want to render along with the Title of the
entity the `type` and the `excerpt`.

In the example below we're using the `extra` property to carry the `type` and the `excerpt` of the entity.

The `convertEntitiesToContrlOptions` will create the `ControlOption` with the `extra` property containing the post type
and the post excerpt.

```typescript
searchEntities: async (
	phrase,
	postType,
	queryArguments
) => {
	const postsEntities = await searchEntities(
		'post',
		postType,
		phrase,
		{...queryArguments, fields: ['title', 'id', 'type', 'excerpt']}
	);

	return convertEntitiesToControlOptions(
		postsEntities,
		'title',
		'id',

		// Extra properties
		'type',
		'excerpt'
	);
}
```

Then you can access the extra properties within your component since the `CompositeEntitiesByKind` will pass
the `entities` of type `Kensaku.BaseControl< E >` to the children.

```jsx
(entities: Kensaku.BaseControl< E >) => {
    return <MyComponent {...entities} />;
}
```

And internally of your `MyComponent` you can access the `extra` property of the `ControlOption`:

```jsx
function MyComponent(entities: Options<V>): JSX.Element {
    return (
        <div>
            {entities.options.map((option) => (
                <label>
                    <input value={option.value} />
                    <span>{option.label}</span>
                    <span>
                        <span>{option.extra.get('type')}</span>
                        <span>{option.extra.get('excerpt')}</span>
                    </span>
                </label>
            ))}
        </div>
    );
}
```
