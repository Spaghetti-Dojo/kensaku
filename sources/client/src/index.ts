export * from './api/search-entities';

export * from './components/composite-entities-by-kind';
export * from './components/plural-select-control';
export * from './components/radio-control';
export * from './components/search-control';
export * from './components/singular-select-control';
export * from './components/toggle-control';

export * from './hooks/use-search';
export * from './hooks/use-entity-records';
export * from './hooks/use-query-viewable-post-types';
export * from './hooks/use-query-viewable-taxonomies';

export * from './utils/convert-entities-to-control-options';

export * from './vo/set';
