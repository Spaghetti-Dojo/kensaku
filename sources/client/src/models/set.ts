/**
 * External dependencies
 */
import { isEqual as _isEqual } from 'lodash';

export class Set< T > {
	readonly #data: ReadonlyArray< T >;

	public constructor( data: ReadonlyArray< T > = [] ) {
		this.#data = this.ensureUnique( data );
	}

	public add( value: T ): Set< T > {
		if ( this.has( value ) ) {
			return this;
		}

		return new Set( [ ...this.#data, value ] );
	}

	public delete( value: T ): Set< T > {
		if ( ! this.has( value ) ) {
			return this;
		}

		return new Set(
			this.#data.filter( ( item ) => ! this.isEqual( item, value ) )
		);
	}

	public has( value: T ): boolean {
		return this._has( value, this.#data );
	}

	public map< R = T >( fn: ( value: T ) => R ): Set< R > {
		return new Set( this.#data.map( fn ) );
	}

	public toArray(): ReadonlyArray< T > {
		return Object.freeze( [ ...this.#data ] );
	}

	public forEach( fn: ( value: T ) => void ): void {
		this.#data.forEach( fn );
	}

	public length(): number {
		return this.#data.length;
	}

	public concat( set: Set< T > ): Set< T > {
		return new Set( [ ...this.#data, ...set.toArray() ] );
	}

	public filter( fn: ( value: T ) => boolean ): Set< T > {
		return new Set( this.#data.filter( fn ) );
	}

	public find( fn: ( value: T ) => boolean ): T | undefined {
		return this.#data.slice( 0 ).find( fn );
	}

	public first(): T | undefined {
		return this.#data.slice( 0 )[ 0 ];
	}

	public last(): T | undefined {
		return this.#data.slice( -1 )[ 0 ];
	}

	public copy( start: number, end: number ): Set< T > {
		return new Set( this.#data.slice( start, end ) );
	}

	public equals( set: Set< T > ): boolean {
		if ( this.length() !== set.length() ) {
			return false;
		}

		if ( this === set ) {
			return true;
		}

		for ( const value of this ) {
			if ( ! set.has( value ) ) {
				return false;
			}
		}

		return true;
	}

	public *[ Symbol.iterator ]() {
		for ( const value of this.#data ) {
			yield value;
		}
	}

	private isEqual( a: unknown, b: unknown ): boolean {
		return _isEqual( a, b );
	}

	private _has( value: T, data: ReadonlyArray< T > ): boolean {
		return data.some( ( current ) => this.isEqual( current, value ) );
	}

	private ensureUnique( data: ReadonlyArray< T > ): ReadonlyArray< T > {
		const accumulator: Array< T > = [];
		return data.reduce( ( acc, value ) => {
			if ( ! this._has( value, acc ) ) {
				acc.push( value );
			}

			return acc;
		}, accumulator );
	}
}
