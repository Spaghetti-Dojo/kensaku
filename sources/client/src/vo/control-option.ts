/**
 * External dependencies
 */
import EntitiesSearch from '@types';

/**
 * Internal dependencies
 */
import { assert } from '../utils/assert';

export class ControlOption< V > implements EntitiesSearch.ControlOption< V > {
	public readonly label: string;
	public readonly value: V;

	public constructor( label: string, value: V ) {
		assert(
			label !== '',
			'ControlOption: Label must be a non empty string.'
		);
		assert(
			value !== '',
			'ControlOption: Value must be a non empty string.'
		);

		this.label = label;
		this.value = value;
	}
}
