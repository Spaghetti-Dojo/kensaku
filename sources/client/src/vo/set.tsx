export class Set<T> {
	readonly #data: ReadonlyArray<T>;

	public constructor(data: Array<T> = []) {
		this.#data = data;
	}

	public add(value: T): Set<T> {
		if (this.has(value)) {
			return this;
		}

		return new Set([...this.#data, value]);
	}

	public delete(value: T): Set<T> {
		if (!this.has(value)) {
			return this;
		}

		return new Set(this.#data.filter((item) => item !== value));
	}

	public has(value: T): boolean {
		return this.#data.includes(value);
	}

	public map<R = T>(fn: (value: T) => R): Set<R> {
		return new Set(this.#data.map(fn));
	}

	public toArray(): Array<T> {
		return [...this.#data];
	}

	public forEach(fn: (value: T) => void): void {
		this.#data.forEach(fn);
	}

	public length(): number {
		return this.#data.length;
	}

	public concat(set: Set<T>): Set<T> {
		return new Set([...this.#data, ...set.toArray()]);
	}

	public filter(fn: (value: T) => boolean): Set<T> {
		return new Set(this.#data.filter(fn));
	}

	public first(): T | undefined {
		return this.#data.slice(0)[0];
	}

	public last(): T | undefined {
		return this.#data.slice(-1)[0];
	}

	public copy(start: number, end: number): Set<T> {
		return new Set(this.#data.slice(start, end));
	}

	public equals(set: Set<T>): boolean {
		if (this.length() !== set.length()) {
			return false;
		}

		if (this === set) {
			return true;
		}

		for (const value of this) {
			if (!set.has(value)) {
				return false;
			}
		}

		return true;
	}

	public *[Symbol.iterator]() {
		for (const value of this.#data) {
			yield value;
		}
	}
}
