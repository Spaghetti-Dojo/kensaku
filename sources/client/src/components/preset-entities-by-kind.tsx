/**
 * External dependencies
 */
import Kensaku from '@types';
import React, { JSX } from 'react';
import classnames from 'classnames';

/**
 * WordPress dependencies
 */
import { createHigherOrderComponent } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import { CompositeEntitiesByKind } from './composite-entities-by-kind';
import { SearchControl } from './search-control';
import { Set } from '../models/set';

type EntitiesValue = Kensaku.Value;
type Entities = Kensaku.Entities< EntitiesValue >;
type KindValue = Kensaku.Kind< string > | string;

type EntitiesFinder = (
	phrase: string,
	kind: Kensaku.Kind< string >,
	queryArguments?: Kensaku.QueryArguments
) => Promise< Set< Kensaku.ControlOption< EntitiesValue > > >;

type EntitiesComponent = React.ComponentType<
	Kensaku.BaseControl< EntitiesValue >
>;
type KindComponent = React.ComponentType< Kensaku.BaseControl< KindValue > >;

type PublicComponentProps = {
	entitiesFinder: EntitiesFinder;
	className?: string;
	entities: Entities;
	onChangeEntities: ( values: Entities ) => void;
	entitiesComponent: EntitiesComponent;
	kind: KindValue;
	kindOptions: Kensaku.Options< string >;
	onChangeKind: ( values: KindValue ) => void;
	kindComponent: KindComponent;
	entitiesFields?: Kensaku.QueryArguments[ 'fields' ];
};

interface PrivateComponentProps
	extends Pick<
		Kensaku.CompositeEntitiesKinds< EntitiesValue, string >,
		'kind' | 'entities'
	> {
	className?: string;
	entitiesFinder: EntitiesFinder;
	kindComponent: KindComponent;
	entitiesComponent: EntitiesComponent;
}

function PrivateComponent( props: PrivateComponentProps ): JSX.Element {
	const className = classnames( 'kensaku-preset-entities-by-kind', {
		// @ts-ignore
		[ props.className ]: !! props.className,
	} );

	return (
		<div className={ className }>
			<CompositeEntitiesByKind
				entities={ props.entities }
				kind={ props.kind }
				searchEntities={ props.entitiesFinder }
			>
				{ ( _entities, _kind, search ) => (
					<>
						<props.kindComponent { ..._kind } />
						<SearchControl onChange={ search } />
						<props.entitiesComponent { ..._entities } />
					</>
				) }
			</CompositeEntitiesByKind>
		</div>
	);
}

const withDataBound = createHigherOrderComponent<
	React.ComponentType< PrivateComponentProps >,
	React.ComponentType< PublicComponentProps >
>(
	( Component ) => ( props ) => {
		const {
			kind,
			entities,
			onChangeEntities,
			kindOptions,
			onChangeKind,
			entitiesFields,
			kindComponent,
			entitiesComponent,
			entitiesFinder,
			..._props
		} = props;

		const kindValue = narrowKindValue( props.kind );

		const _entities = {
			value: entities,
			onChange: onChangeEntities,
		};

		const _kind = {
			value: kindValue,
			options: kindOptions,
			onChange: onChangeKind,
		};

		const _entitiesFinder = entitiesFinderWithExtraFields(
			entitiesFinder,
			entitiesFields
		);

		return (
			<Component
				entities={ _entities }
				kind={ _kind }
				entitiesFinder={ _entitiesFinder }
				kindComponent={ kindComponent }
				entitiesComponent={ entitiesComponent }
				{ ..._props }
			/>
		);
	},
	'withDataBound'
);

function entitiesFinderWithExtraFields(
	entitiesFinder: EntitiesFinder,
	entitiesFields?: Kensaku.QueryArguments[ 'fields' ]
): EntitiesFinder {
	return (
		phrase: string,
		kind: Kensaku.Kind< string >,
		queryArguments?: Kensaku.QueryArguments
	) =>
		entitiesFinder( phrase, kind, {
			...queryArguments,
			fields: [
				...( queryArguments?.fields ?? [ 'title', 'id' ] ),
				...( entitiesFields ?? [] ),
			],
		} );
}

function narrowKindValue( value: KindValue ): Kensaku.Kind< string > {
	return typeof value === 'string' ? new Set( [ value ] ) : value;
}

export const PresetEntitiesByKind = withDataBound( PrivateComponent );
