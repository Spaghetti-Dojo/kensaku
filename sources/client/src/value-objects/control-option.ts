/**
 * External dependencies
 */
import Kensaku from '@types';

/**
 * Internal dependencies
 */
import { assert } from '../utils/assert';
import { ImmutableRecord } from '../models/immutable-record';

export class ControlOption< V > implements Kensaku.EnrichedControlOption< V > {
	public readonly label: string;
	public readonly value: V;
	public readonly extra: Kensaku.Record< unknown >;

	public constructor(
		label: string,
		value: V,
		extra: Kensaku.Record< unknown > = new ImmutableRecord()
	) {
		assert(
			label !== '',
			'ControlOption: Label must be a non empty string.'
		);
		assert(
			value !== '',
			'ControlOption: Value must be a non empty string.'
		);

		this.label = label;
		this.value = value;
		this.extra = extra;
	}
}
