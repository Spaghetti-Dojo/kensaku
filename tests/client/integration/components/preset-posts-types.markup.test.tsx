/**
 * External dependencies
 */
import React from 'react';
import { describe, expect, it, jest } from '@jest/globals';
import { render, screen, act } from '@testing-library/react';

/**
 * Internal dependencies
 */
import { Set } from '../../../../sources/client/src/models/set';
import { createSearchEntitiesOptions } from '../../../../sources/client/src';
import { ControlOption } from '../../../../sources/client/src/value-objects/control-option';
import { PresetEntitiesByKind } from '../../../../sources/client/src/components/preset-entities-by-kind';

describe( 'Preset Entities By Kind', () => {
	it( 'Should render the CompositeEntitiesByKind component with the appropriate configuration', async () => {
		const entitiesFinder =
			jest.fn<
				ReturnType< typeof createSearchEntitiesOptions< string > >
			>();

		const props = {
			entitiesFinder,
			entities: new Set( [ 1, 2, 3 ] ),
			onChangeEntities: jest.fn(),
			entitiesComponent: () => <div>EntitiesComponent</div>,
			kind: 'post',
			kindOptions: stubControlOptionsSet(),
			onChangeKind: jest.fn(),
			kindComponent: () => <div>KindComponent</div>,
			className: 'extra-class-name',
		};

		const rendered = await act( () =>
			render( <PresetEntitiesByKind { ...props } /> )
		);

		expect( rendered.asFragment() ).toMatchSnapshot();
	} );

	it( 'Expect Kind Component to be rendered as first component', async () => {
		const entitiesFinder =
			jest.fn<
				ReturnType< typeof createSearchEntitiesOptions< string > >
			>();

		const props = {
			entitiesFinder,
			entities: new Set( [ 1, 2, 3 ] ),
			onChangeEntities: jest.fn(),
			entitiesComponent: () => (
				<div data-testid="entities-component">EntitiesComponent</div>
			),
			kind: 'post',
			kindOptions: stubControlOptionsSet(),
			onChangeKind: jest.fn(),
			kindComponent: () => (
				<div data-testid="kind-component">KindComponent</div>
			),
		};

		const rendered = await act( () =>
			render( <PresetEntitiesByKind { ...props } /> )
		);

		const kindComponent = screen.getByTestId( 'kind-component' );
		const firstComponent = rendered.container.querySelector(
			'.kensaku-preset-entities-by-kind'
		);

		expect( kindComponent === firstComponent?.firstElementChild ).toEqual(
			true
		);
	} );

	it( 'Expect Entities Component to be rendered as last component', async () => {
		const entitiesFinder =
			jest.fn<
				ReturnType< typeof createSearchEntitiesOptions< string > >
			>();

		const props = {
			entitiesFinder,
			entities: new Set( [ 1, 2, 3 ] ),
			onChangeEntities: jest.fn(),
			entitiesComponent: () => (
				<div data-testid="entities-component">EntitiesComponent</div>
			),
			kind: 'post',
			kindOptions: stubControlOptionsSet(),
			onChangeKind: jest.fn(),
			kindComponent: () => (
				<div data-testid="kind-component">KindComponent</div>
			),
		};

		const rendered = await act( () =>
			render( <PresetEntitiesByKind { ...props } /> )
		);

		const kindComponent = screen.getByTestId( 'entities-component' );
		const firstComponent = rendered.container.querySelector(
			'.kensaku-preset-entities-by-kind'
		);

		expect( kindComponent === firstComponent?.lastElementChild ).toEqual(
			true
		);
	} );

	it( 'Allows extra className property', async () => {
		const entitiesFinder =
			jest.fn<
				ReturnType< typeof createSearchEntitiesOptions< string > >
			>();

		const props = {
			entitiesFinder,
			entities: new Set( [ 1, 2, 3 ] ),
			onChangeEntities: jest.fn(),
			entitiesComponent: () => (
				<div data-testid="entities-component">EntitiesComponent</div>
			),
			kind: 'post',
			kindOptions: stubControlOptionsSet(),
			onChangeKind: jest.fn(),
			kindComponent: () => (
				<div data-testid="kind-component">KindComponent</div>
			),
			className: 'extra-class',
		};

		const rendered = await act( () =>
			render( <PresetEntitiesByKind { ...props } /> )
		);

		expect(
			rendered.container
				.querySelector( '.kensaku-preset-entities-by-kind' )
				?.classList.contains( 'extra-class' )
		).toEqual( true );
	} );
} );

function stubControlOptionsSet() {
	return new Set( [
		new ControlOption( 'Post', 'post' ),
		new ControlOption( 'Page', 'page' ),
	] );
}
