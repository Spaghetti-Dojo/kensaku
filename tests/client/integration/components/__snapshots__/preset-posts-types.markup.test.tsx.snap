// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preset Entities By Kind Should render the CompositeEntitiesByKind component with the appropriate configuration 1`] = `
<DocumentFragment>
  <div
    class="kensaku-preset-entities-by-kind extra-class-name"
  >
    <div>
      KindComponent
    </div>
    <div
      class="kensaku-search-control"
    >
      <label
        for=":r0:"
      >
        <span
          class="kensaku-search-control__label"
        >
          Search
        </span>
        <input
          class="kensaku-search-control__input"
          id=":r0:"
          type="search"
          value=""
        />
      </label>
    </div>
    <div>
      EntitiesComponent
    </div>
  </div>
</DocumentFragment>
`;
